### PR TITLE
Create conversions package

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Moves much of the code around creation of storage variants out of the **astmodel** package and into a package dedicated to generation of conversions.

~Requires PR #1495 to be merged first.~ PR #1495 has been merged.

**Special notes for your reviewer**:

The PR is almost entirely moving files and renaming types (see below for the exception). However, not every file rename has been detected and correctly shown in the PR, so in some cases we see new-file + file-deleted.

The one functional change is that functions are named for the type being converted, not for the version. This can be seen in the changes to the (relocated) golden files, where `ConvertToVNext()` now becomes `AssignPropertiesToPerson()`. In the actuall generated code, where the versions are date based, the change will look like `ConvertToV20180501()` becoming `AssignPropertiesToBatchAccount()`.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: hack/generator/pkg/conversions/property_assignment_function_test.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package conversions
+
+import (
+	"bytes"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
```
Comment (babbageclunk) [inline hack/generator/pkg/conversions/property_assignment_function_test.go:10]: Same

File: hack/generator/pkg/conversions/property_conversion_context.go [hunk 2]
```diff
@@ -3,67 +3,80 @@
...
+// NewStorageConversionContext creates a new instance of a PropertyConversionContext
+func NewStorageConversionContext(types astmodel.Types, idFactory astmodel.IdentifierFactory) *PropertyConversionContext {
+	return &PropertyConversionContext{
 		types:       types,
-		knownLocals: NewKnownLocalsSet(idFactory),
+		knownLocals: astmodel.NewKnownLocalsSet(idFactory),
 	}
 }
 
+//TODO: Remove this?
```
Comment (babbageclunk) [inline hack/generator/pkg/conversions/property_conversion_context.go:30]: So should you? :wink:

File: hack/generator/pkg/conversions/property_conversion_endpoint.go [hunk 3]
```diff
@@ -3,50 +3,56 @@
  * Licensed under the MIT license.
  */
 
-package astmodel
+package conversions
 
 import (
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
```
Comment (babbageclunk) [inline hack/generator/pkg/conversions/property_conversion_endpoint.go:9]: Same here

File: hack/generator/pkg/conversions/property_conversions.go [hunk 4]
```diff
@@ -3,9 +3,10 @@
  * Licensed under the MIT license.
  */
 
-package astmodel
+package conversions
 
 import (
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
```
Comment (babbageclunk) [inline hack/generator/pkg/conversions/property_conversions.go:9]: Same

**Review by babbageclunk — APPROVED** [review]
I like the AssignPropertiesTo/From naming a lot more.

File: hack/generator/pkg/conversions/property_assignment_function_test.go (hunk 1 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/conversions/property_assignment_function_test.go:10]: Fixed.

File: hack/generator/pkg/codegen/storage/storage_type_factory.go [hunk 5]
```diff
@@ -7,6 +7,7 @@ package storage
 
 import (
 	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/conversions"
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/codegen/storage/storage_type_factory.go:10]: Fixed.

File: hack/generator/pkg/conversions/property_assignment_function.go [hunk 6]
```diff
@@ -3,10 +3,11 @@
  * Licensed under the MIT license.
  */
 
-package astmodel
+package conversions
 
 import (
 	"fmt"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply hack/generator/pkg/conversions/property_assignment_function.go:10]: Fixed.

File: hack/generator/pkg/conversions/property_conversion_context.go (hunk 2 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/conversions/property_conversion_context.go:30]: Those methods were introduced as a part of relocating things one by one into the new package, but I'll keep them. Fixed the comments.

File: hack/generator/pkg/conversions/property_conversion_endpoint.go (hunk 3 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/conversions/property_conversion_endpoint.go:9]: Fixed.

File: hack/generator/pkg/conversions/property_conversions.go (hunk 4 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/conversions/property_conversions.go:9]: Fixed.

**Review by babbageclunk — APPROVED** [review]
Nice fix!

**Pull request merged**